python/dgl/node_flow.py

```diff
@@ -0,0 +1,845 @@
+        self._node_frames = [FrameRef(Frame(num_rows=self.layer_size(i))) \
+                             for i in range(self.num_layers)]
+        self._edge_frames = [FrameRef(Frame(num_rows=self.block_size(i))) \
+                             for i in range(self.num_blocks)]
+        # registered functions
+        self._message_funcs = [None] * self.num_blocks
+        self._reduce_funcs = [None] * self.num_blocks
+        self._apply_node_funcs = [None] * self.num_blocks
+        self._apply_edge_funcs = [None] * self.num_blocks
+
+    def _get_layer_id(self, layer_id):
+        if layer_id >= 0:
+            return layer_id
+        else:
+            return self.num_layers + layer_id
+
+    def _get_block_id(self, block_id):
+        if block_id >= 0:
+            return block_id
+        else:
+            return self.num_blocks + block_id
+
+    def _get_node_frame(self, layer_id):
+        return self._node_frames[layer_id]
+
+    def _get_edge_frame(self, flow_id):
+        return self._edge_frames[flow_id]
+
+    @property
+    def num_layers(self):
+        """Get the number of layers.
+
+        Returns
+        -------
+        int
+            the number of layers
+        """
+        return len(self._layer_offsets) - 1
+
+    @property
+    def num_blocks(self):
+        """Get the number of blocks.
+
+        Returns
+        -------
+        int
+            the number of blocks
+        """
+        return self.num_layers - 1
+
+    @property
+    def layers(self):
+        """Return a LayerView of this NodeFlow.
+
+        This is mainly for usage like:
+        * `g.layers[2].data['h']` to get the node features of layer#2.
+        * `g.layers(2)` to get the nodes of layer#2.
+        """
+        return LayerView(self)
+
+    @property
+    def blocks(self):
+        """Return a BlockView of this NodeFlow.
+
+        This is mainly for usage like:
+        * `g.blocks[1,2].data['h']` to get the edge features of blocks from layer#1 to layer#2.
+        * `g.blocks(1, 2)` to get the edge ids of blocks #1->#2.
+        """
+        return BlockView(self)
+
+    def layer_size(self, layer_id):
+        """Return the number of nodes in a specified layer.
+
+        Parameters
+        ----------
+        layer_id : int
+            the specified layer to return the number of nodes.
+        """
+        layer_id = self._get_layer_id(layer_id)
+        return int(self._layer_offsets[layer_id + 1]) - int(self._layer_offsets[layer_id])
+
+    def block_size(self, block_id):
+        """Return the number of edges in a specified block.
+
+        Parameters
+        ----------
+        block_id : int
+            the specified block to return the number of edges.
+        """
+        block_id = self._get_block_id(block_id)
+        return int(self._block_offsets[block_id + 1]) - int(self._block_offsets[block_id])
+
+    def copy_from_parent(self, node_embed_names=ALL, edge_embed_names=ALL):
```

Comment:
Is this the same logic from subgraph? Can we reuse it?

Reply:
Similar, but it's not the same, mainly because of the concept of layers.